grpc-server.go

```diff
@@ -696,5 +699,113 @@ func blockContainsAccounts(block *old_faithful_grpc.BlockResponse, accounts []st
+	ctx := ser.Context()
+
+	startSlot := params.StartSlot
+	endSlot := startSlot + maxSlotsToStream
+
+	if params.EndSlot != nil {
+		endSlot = *params.EndSlot
+	}
+	gsfaReader, _ := multi.getGsfaReadersInEpochDescendingOrderForSlotRange(ctx, startSlot, endSlot)
+
+	for slot := startSlot; slot <= endSlot; slot++ {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+		}
+
+		if err := multi.processSlotTransactions(ctx, ser, slot, params.Filter, gsfaReader); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (multi *MultiEpoch) processSlotTransactions(
+	ctx context.Context,
+	ser old_faithful_grpc.OldFaithful_StreamTransactionsServer,
+	slot uint64, filter *old_faithful_grpc.StreamTransactionsFilter,
+	gsfaReader *gsfa.GsfaReaderMultiepoch,
+) error {
+
+	filterFunc := func(txn *ipldbindcode.Transaction) bool {
+		// fill this out
+
+		return true
+	}
 
+	if filter == nil || len(filter.AccountInclude) == 0 {
+
+		// get block -> not sure which one to use
+		// block -> transactions
+		// Apply filters
+		// Send
+	} else {
+		for _, account := range filter.AccountInclude {
+			pKey := solana.MustPublicKeyFromBase58(account)
+			epochToTxns, err := gsfaReader.Get(
+				ctx,
+				pKey,
+				1000,
+				func(epochNum uint64, oas linkedlog.OffsetAndSizeAndBlocktime) (*ipldbindcode.Transaction, error) {
+					epoch, err := multi.GetEpoch(epochNum)
+					if err != nil {
+						return nil, fmt.Errorf("failed to get epoch %d: %w", epochNum, err)
+					}
+					raw, err := epoch.GetNodeByOffsetAndSize(ctx, nil, &indexes.OffsetAndSize{
+						Offset: oas.Offset,
+						Size:   oas.Size,
+					})
+					if err != nil {
+						return nil, fmt.Errorf("failed to get signature: %w", err)
+					}
+					decoded, err := iplddecoders.DecodeTransaction(raw)
+					if err != nil {
+						return nil, fmt.Errorf("error while decoding transaction from nodex at offset %d: %w", oas.Offset, err)
+					}
+					return decoded, nil
+				},
+			)
+
+			if err != nil {
+				return err
+			}
+
+			for epochNumber, txns := range epochToTxns {
+				epochHandler, err := multi.GetEpoch(epochNumber)
+				if err != nil {
+					return status.Errorf(codes.NotFound, "Epoch %d is not available", epochNumber)
+				}
+				for _, txn := range txns {
+					txResp := new(old_faithful_grpc.TransactionResponse)
+					txResp.Transaction = new(old_faithful_grpc.Transaction)
+					{
+						pos, ok := txn.GetPositionIndex()
+						if ok {
+							txResp.Index = ptrToUint64(uint64(pos))
+							txResp.Transaction.Index = ptrToUint64(uint64(pos))
+						}
+						txResp.Transaction.Transaction, txResp.Transaction.Meta, err = getTransactionAndMetaFromNode(txn, epochHandler.GetDataFrameByCid)
+						if err != nil {
+							return status.Errorf(codes.Internal, "Failed to get transaction: %v", err)
+						}
+						txResp.Slot = uint64(txn.Slot)
+						// What to do for blocktime?
+					}
+
+					// not sure how to apply more filters
```

Comment:
There's three meta structures:
- confirmed_block.TransactionStatusMeta
- metalatest.TransactionStatusMeta
- metaoldest.TransactionStatusMeta

if you have a [TransactionStatusMetaContainer](https://github.com/rpcpool/yellowstone-faithful/blob/2d89056c290c1f0c026ac66deed849daee28f1f5/solana-tx-meta-parsers/parsers.go#L12) you can know which one it is

Reply:
Actually you may not even need the TransactionStatusMetaContainer, it seems you can just use the [ParseAnyTransactionStatusMeta](https://github.com/rpcpool/yellowstone-faithful/blob/85b6d2c63eea7ff9442cd34ebb02dfd2b50e0da2/solana-tx-meta-parsers/parsers.go#L85).

For err, Slavomir has a function to get the error:
https://github.com/rpcpool/yellowstone-faithful/blob/85b6d2c63eea7ff9442cd34ebb02dfd2b50e0da2/err.go#L10-L11

So that should be all you need for filtering error transactions.

Reply:
For votes I found this definition:
https://github.com/rpcpool/solana-private/blob/f6e0de9a7c76e18a5ad11752806308860f77bd20/sdk/src/simple_vote_transaction_checker.rs#L4

We can replicate this definition in a new utility ufnction similar to getErr